x-pack/plugins/apm/public/components/shared/FilterBar/ServiceEnvFilter.tsx

```diff
@@ -0,0 +1,94 @@
+import { EuiFilterGroup, EuiFilterButton, EuiSelect } from '@elastic/eui';
+import React, { useEffect, useRef, useState } from 'react';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+function without<T>(obj: T, key: string): T {
+  const result: { [objKey: string]: unknown } = Object.assign({}, obj);
+  delete result[key];
+  return result as T;
+}
+
+export const ServiceEnvFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { start, end, kuery, environment } = urlParams;
+  const { data = [] } = useFetcher(
+    () => loadServiceEnvironments({ start, end, kuery, environment: '' }),
+    [start, end, kuery]
+  );
+  const availableEnvironments = ['Any', ...data, 'Not set'];
+  const [selectedValue, setSelectedValue] = useState(availableEnvironments[0]);
+  const isPostEffect = useRef(false);
+
+  function updateEnvironmentUrl(environment?: string) {
+    if (!environment || environment === availableEnvironments[0]) {
+      history.push({
+        ...location,
+        search: fromQuery(without(toQuery(location.search), 'environment'))
+      });
+    } else if (
+      environment === availableEnvironments[availableEnvironments.length - 1]
+    ) {
+      history.push({
+        ...location,
+        search: fromQuery({
+          ...toQuery(location.search),
+          environment: 'Null'
```

Comment:
why `Null` with uppercase - and as a string?

Reply:
I understand the meaning of "Null" now. Since it's reused you should make it a shared constant and add a comment to it.

Reply:
this value is arbitrary, since it will ultimately be transformed to the appropriate ES query as needed. I just chose a value that looked nice in the URL, but wasn't `'null'` since that string is a special case.

Reply:
That makes sense. It's just hard to figure out if it's a mistake and if not why it is like that. But either way, I'm voting for postponing this until we get the other things right.